File: 121. Best Time to Buy and Sell Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        
        //brute force approach - o(n^2)
        /*
        int max =0;
        for(int i=0;i<prices.length-1;i++)
        {
            for(int j=i+1;j<prices.length;j++)
            {
                max = Math.max(max, prices[j] - prices[i]);
            }
        }
        return max;

        */

        //single pass approach
        // logic - Need to find the maximum difference between the prices[i] and prices[j] where i<j
        int lowest_yet = Integer.MAX_VALUE;
        int highest_yet = 0;
        int maxdiff = 0;
        for(int i=0;i<prices.length;i++)
        {
            if(prices[i]<lowest_yet)
            {
                lowest_yet = Math.min(lowest_yet, prices[i]); 
            }
            else 
            {
                
                maxdiff = Math.max(maxdiff, prices[i]-lowest_yet);
            }
        

        }
        return maxdiff;

    }

    

}

File: python_codes/121. Best Time to Buy and Sell Stock.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        lowest_yet =  float('inf')
        highest_yet = 0
        maxdiff = 0

        for p in prices:
            if p < lowest_yet:
                lowest_yet = min(p,lowest_yet)
            else:
                diff = p-lowest_yet

                maxdiff = max(diff, maxdiff)
        
        return maxdiff


if __name__ == "__main__":
    sol = Solution()

    test_cases = [
        ([7, 1, 5, 3, 6, 4], 5),
        ([7, 6, 4, 3, 1], 0),
        ([1, 2, 3, 4, 5], 4),
        ([3, 3, 3, 3], 0),
        ([2], 0),
        ([], 0),
        ([1, 7, 2, 10, 4, 8], 9),
        ([5, 1, 5, 3, 6, 4], 5),
    ]

    for prices, expected in test_cases:
        result = sol.maxProfit(prices)
        assert result == expected, f"maxProfit({prices}) == {result}, expected {expected}"
        print(f"Passed: maxProfit({prices}) == {expected}")

    print("All tests passed!")